products/accredited-representation-management/technical-documentation/appoint-a-representative/poa_request_proposal.md

```diff
@@ -0,0 +1,160 @@
+# Storage Proposal for Power of Attorney Requests
+This proposal only covers the storing of data for Power of Attorney Requests.
+
+## Background
+The Office of the Chief Technology Officer (OCTO) is seeking to migrate Veteran and Power of Attorney (PoA) experiences from the legacy Stakeholder Enterprise Platform (SEP) product and eBenefits platform to VA.gov.
+
+One aspect of this experience is the Power of Attorney (POA) Request, a data object that exists to enable accredited representatives accepting or denying requests to act as a Power of Attorney (Forms 21-22 and 21-22a) on behalf of Veterans. The data object bridges the gap from the Veteran facing form fill experience (Appoint a Representative) to establishing the Power of Attorney relationship in systems of record and updating representative permissions in those systems (Lighthouse auto-establishment APIs). [See the flow chart](#data-flow)
+
+## Problem
+In order to create a functional user interface for accredited representatives ([see the sample UI](#sample-ui)), the POA Request objects must support reasonably efficient searching or filtering on the following Veteran attributes: first name, last name, city, state, zip code, and consent limits. The front end must also be able to display information related to change of address authorization and health record access limitations.
+
+There is not an existing pattern or solution in vets-api for storing Veteran PII that is secure and efficiently searchable for a many row result.
+
+## Solution
+Code below is for illustrative purposes. These snippets are not likely to be the final implementation.
+
+### Column level encryption
+The [blind index gem](https://github.com/ankane/blind_index) allows for exact matching queries on encrypted columns. This will be used on fields that users will search or filter by. Existing KMS encryption will be used on the fields that are for display purposes only.
+
+#### The Model
+```ruby
+class PowerOfAttorneyRequest < ApplicationRecord
+  has_kms_key
+  has_encrypted :consent_data, key: :kms_key, **lockbox_options
+  blind_index :first_name, :last_name, :city, :state_code, :zip_code
```

Comment:
I am skeptical that the product needs to support "VSO can filter by all 'Adam's". I believe product does want to load data for a particular user. For this there might be a part of the app that requires a sufficient number of inputs to uniquely identify a Veteran/claimant that a VSO could enter that we could get back to the user from using MPI. SEP's version of this feature is described in SEP FAQs [here](https://www.knowva.ebenefits.va.gov/system/templates/selfservice/va_ssnew/help/customer/locale/en-US/portal/554400000001018/content/554400000008534/61-Search-for-a-Veteran).

Reply:
For storage purposes then, switching first and last to KMS would be good? We lose the easy search/filter but would still be able to display the name without an MPI call for each Veteran.

Reply:
@nihil2501, is your suggestion to use `mpi_service.find_profile_by_attributes(first_name:, last_name:, birth_date:, ssn:)` instead of blind indexing to find a single record?

Reply:
I think there are two features of the app that concern the details of a single Veteran. One is the POA request detail view. For this we would load the POA request by ID, which in turn is associated to a user that has an ICN which we could use to retrieve Veteran details from MPI.

The other is like the one linked in SEP where the VSO enters in fields including SSN and DOB that I assume they have on hand in this flow. A query of MPI with those fields would yield an ICN that we could use to find a corresponding Veteran user's POA requests. I don't think this feature has been mocked up yet though. Also I'm curious about the same question you have about the required field set in the SEP instructions I linked.